clash-prelude/src/Clash/Explicit/BlockRam.hs

```diff
@@ -1094,3 +1108,278 @@ readNew clk rst en ram rdAddr wrM = mux wasSame wasWritten $ ram rdAddr wrM
+
+isRamWrite :: RamOp n a -> Bool
+isRamWrite (RamWrite {}) = True
+isRamWrite _             = False
+
+ramOpWriteVal :: RamOp n a -> Maybe a
+ramOpWriteVal (RamWrite _ val) = Just val
+ramOpWriteVal _                = Nothing
+
+isOp :: RamOp n a -> Bool
+isOp NoOp = False
+isOp _    = True
+
+
+-- | Produces vendor-agnostic HDL that will be inferred as a true, dual port
+-- block ram. Any values that's being written on a particular port is also the
+-- value that will be read on that port, i.e. the same-port read/write behavior
+-- is: WriteFirst. For mixed port read/write, when both ports have the same
+-- address, when there is a write on the port A, the output of port B is
+-- undefined, and visa versa. Implicitly clocked version is
+-- `Clash.Prelude.BlockRam.trueDualPortBlockRam`
+trueDualPortBlockRam ::
+  forall nAddrs domA domB a .
+  ( HasCallStack
+  , KnownNat nAddrs
+  , KnownDomain domA
+  , KnownDomain domB
+  , NFDataX a
+  )
+  => Clock domA
+  -- ^ Clock for port A
+  -> Clock domB
+  -- ^ Clock for port B
+  -> Signal domA (RamOp nAddrs a)
+  -- ^ ram operation for port A
+  -> Signal domB (RamOp nAddrs a)
+  -- ^ ram operation for port B
+  -> (Signal domA a, Signal domB a)
+  -- ^ Outputs data on /next/ cycle. When writing, the data written
+  -- will be echoed. When reading, the read data is returned.
+
+{-# INLINE trueDualPortBlockRam #-}
+trueDualPortBlockRam = \clkA clkB opA opB ->
+  trueDualPortBlockRamWrapper
+    clkA (isOp <$> opA) (isRamWrite <$> opA) (ramOpAddr <$> opA) (fromJustX . ramOpWriteVal <$> opA)
+    clkB (isOp <$> opB) (isRamWrite <$> opB) (ramOpAddr <$> opB) (fromJustX . ramOpWriteVal <$> opB)
+
+toMaybeX :: a -> MaybeX a
+toMaybeX a =
+  case isX a of
+    Left _ -> IsX
+    Right _ -> IsDefined a
+
+data MaybeX a = IsX | IsDefined !a
+
+data Conflict = Conflict
+  { cfWrite :: !(MaybeX Bool)
+  , cfAddress :: !(MaybeX Int) }
+
+instance Semigroup Conflict where
+  (<>) = mergeConflicts
+
+-- | "Stronger" conflict wins:
+--
+--   * Writes over read
+--   * Undefineds over anything
+--
+mergeConflicts :: Conflict -> Conflict -> Conflict
+mergeConflicts conflict1 conflict2 = Conflict
+  { cfWrite = mergeWrite (cfWrite conflict1) (cfWrite conflict2)
+  , cfAddress = mergeAddress (cfAddress conflict1) (cfAddress conflict2) }
+ where
+  mergeX _ IsX _ = IsX
+  mergeX _ _ IsX = IsX
+  mergeX f (IsDefined a) (IsDefined b) = IsDefined (f a b)
+
+  mergeWrite a b = mergeX (||) a b
+  mergeAddress a b = mergeX const a b
+
+trueDualPortBlockRamWrapper clkA enA weA addrA datA clkB enB weB addrB datB =
+  trueDualPortBlockRam# clkA enA weA addrA datA clkB enB weB addrB datB
+{-# NOINLINE trueDualPortBlockRamWrapper #-}
+
+-- | Primitive of 'trueDualPortBlockRam'.
+trueDualPortBlockRam#, trueDualPortBlockRamWrapper ::
+  forall nAddrs domA domB a .
+  ( HasCallStack
+  , KnownNat nAddrs
+  , KnownDomain domA
+  , KnownDomain domB
+  , NFDataX a
+  , (a ~ a)   -- BitPack a -- temp hack to avoid renumbering ~ARGs
```

Comment:
Is it intentional to leave a temp hack in for 1.6.0?

Reply:
Definitely not.